mono/metadata/threads.c

```diff
@@ -4738,28 +4733,31 @@ mono_thread_execute_interruption (void)
 		ERROR_DECL (error);
 		exc = mono_exception_new_thread_interrupted (error);
 		mono_error_assert_ok (error); // FIXME
+		fexc = TRUE;
 	}
 exit:
 	if (unlock)
 		unlock_thread_handle (thread);
 
-	HANDLE_FUNCTION_RETURN_REF (MonoException, exc);
+	if (fexc)
+		MONO_HANDLE_ASSIGN (*pexc, exc);
+
+	return fexc;
 }
 
 static void
 mono_thread_execute_interruption_void (void)
 {
-	(void)mono_thread_execute_interruption ();
+	(void)mono_thread_execute_interruption (NULL);
 }
 
 static MonoException*
 mono_thread_execute_interruption_ptr (void)
 {
 	HANDLE_FUNCTION_ENTER ();
-
-	MonoExceptionHandle exc = mono_thread_execute_interruption ();
-
-	HANDLE_FUNCTION_RETURN_OBJ (exc);
+	MonoExceptionHandle exc;
```

Comment:
Here too: exc should be allocated by the caller.

Reply:
Yeah. I really like the model where we don't allocate these so aggresively, quite inefficient, and a bit nervous-makes-me like in thread shutdown. :(